crates/red_knot_python_semantic/src/types.rs

```diff
@@ -151,6 +165,73 @@ pub(crate) fn definitions_ty<'db>(
     }
 }
 
+/// Build a declared type from a [`DeclarationsIterator`].
+///
+/// Will return a union if there is more than one declaration, or at least one plus the possibility
+/// of undeclared.
+///
+/// If undeclared is a possibility, `Unknown` type will be part of the return type.
+///
+/// # Panics
+/// Will panic if there are no declarations and no possibility of undeclared. This is a logic
+/// error, as any symbol with zero live declarations clearly may (must!) be undeclared.
+fn declarations_ty<'db>(
+    db: &'db dyn Db,
+    declarations: DeclarationsIterator<'_, 'db>,
+) -> DeclaredType<'db> {
+    let may_be_undeclared = declarations.may_be_undeclared();
+    let decl_types = declarations.map(|declaration| declaration_ty(db, declaration));
+
+    let mut all_types = (if may_be_undeclared {
+        Some(Type::Unknown)
+    } else {
+        None
+    })
+    .into_iter()
+    .chain(decl_types);
+
+    let first = all_types.next().expect(
+        "declarations_ty must not be called with zero declarations and no may-be-undeclared.",
+    );
+
+    let mut conflicting: Vec<Type<'db>> = vec![];
+    let declared_ty = if let Some(second) = all_types.next() {
+        let mut builder = UnionBuilder::new(db).add(first);
+        for other in [second].into_iter().chain(all_types) {
+            if !first.is_equivalent_to(db, other) {
+                conflicting.push(other);
+            }
+            builder = builder.add(other);
+        }
+        builder.build()
+    } else {
+        first
+    };
+    DeclaredType {
+        declared_ty,
+        conflicting: if conflicting.is_empty() {
+            TypeList(conflicting.into())
```

Comment:
```suggestion
            TypeList([].into())
```

Reply:
I would have to do some more research but I think an empty `Box<[]>` always allocates (unlike an empty Vec). We may want to use an `Option<Box<[]>>` for `conflicting` for that reason.

Reply:
An other alternative is that this function returns a `Result<Type, (Type, Box<[Type])>>` which woudl also force callers to explicitly handle the error case.

Reply:
Yeah, verified that `Box<[]>` still allocates, so I'll rework this to use `Option`.

Not sure about using `Result` -- it feels odd to me to duplicate the actual return value in both variants, and it's also not clear that all callers _should_ be forced to explicitly handle the error case. In the case of resolving an import (`symbol_ty_by_id`) we want to ignore conflicting declarations and just use the unioned result.

Reply:
In the end I went with a `Result` -- even where we want to ignore conflicting types, it's probably better to be explicit about it.